sql: panic: cannot reset if we've already sent results for this txn

panic while executing "SHOW KV TRACE FOR DROP USER IF EXISTS TIME , REPEATABLE , VARCHAR , ident , FILTER , NAME": cannot reset if we've already sent results for this txn

https://teamcity.cockroachdb.com/viewLog.html?buildId=320932&buildTypeId=Cockroach_Nightlies_RandomSyntaxTests

Found with RSG.

Repro:

```
root@:26257/> begin; SELECT generate_series(1, 10000); select crdb_internal.force_retry('1s'); end;
driver: bad connection
connection lost; opening new connection: all session settings will be lost
```

```
E170814 18:39:58.382124 263 sql/session.go:1128  [client=[::1]:62028,user=root,n1] a SQL panic has occurred!
E170814 18:39:58.382238 263 util/log/crash_reporting.go:81  [n1] a panic has occurred!
panic: cannot reset if we've already sent results for this txn [recovered]
        panic: panic while executing "begin; SELECT generate_series(1, 10000); select crdb_internal.force_retry('1s'); end;": cannot reset if we've already sent results for this txn [recovered]
        panic: panic while executing "begin; SELECT generate_series(1, 10000); select crdb_internal.force_retry('1s'); end;": cannot reset if we've already sent results for this txn [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x4570507]

goroutine 263 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.sendCrashReport(0x77a41e0, 0xc420390f30, 0x0, 0x0, 0x54e14c0, 0xc420bb7820, 0x4)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:150 +0x37
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x77a41e0, 0xc420390f30, 0x0, 0x0, 0x559b2a0, 0xc420bc6fc0, 0x1)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:100 +0x320
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420510000, 0x77a41e0, 0xc420390f30)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:179 +0xab
panic(0x559b2a0, 0xc420bc6fc0)
        /usr/local/Cellar/go/1.8.3/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc420abe000, 0x77a41e0, 0xc420231290)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:416 +0xa1
panic(0x559b2a0, 0xc420bc6fc0)
        /usr/local/Cellar/go/1.8.3/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc420c56000, 0x575913d, 0x9, 0xc42065702c, 0x55)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1139 +0x28d
panic(0x54e14c0, 0xc420bb7750)
        /usr/local/Cellar/go/1.8.3/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).Reset(0xc420abe000, 0x77a45c0, 0xc420bc86c0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:1138 +0xad
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed.func1(0x77a45c0, 0xc420bc86c0, 0xc420e96420, 0xc420aca998, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:857 +0x1e3
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc420e96420, 0x77a45c0, 0xc420bc86c0, 0x1, 0xc420402600, 0x1614f135, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:722 +0xa8
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed(0xc4203fe480, 0xc420c56000, 0xc42000aa00, 0x4, 0x4, 0x0, 0x0, 0xc420021900, 0x6274500)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:882 +0x47c
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4203fe480, 0xc420c56000, 0xc42065702c, 0x55, 0x0, 0x0, 0x55, 0xfd4)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:767 +0x482
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4203fe480, 0xc420c56000, 0xc42065702c, 0x55, 0x0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:649 +0x129
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc420abe000, 0xc420abe028, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:578 +0x148
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420abe000, 0x77a41e0, 0xc420231290, 0xc420aca450, 0x5400, 0xc42001c568, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:496 +0x8bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42001c400, 0x77a45c0, 0xc420bc81c0, 0x629a8e0, 0xc420bcc0a0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x96c
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x629a8e0, 0xc420bcc0a0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/server/server.go:733 +0x149
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc420510000, 0x77a41e0, 0xc420390f30, 0xc4206fe000, 0x629a8e0, 0xc420bcc0a0, 0xc420d4e000)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:140 +0xa9
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0x267
```

--- a/pkg/internal/client/db_test.go
+++ b/pkg/internal/client/db_test.go
@@ -407,6 +407,7 @@ func TestCommonMethods(t *testing.T) {
 		{txnType, "IsSerializableRestart"}:           {},
 		{txnType, "NewBatch"}:                        {},
 		{txnType, "Exec"}:                            {},
+		{txnType, "PrepareForRetry"}:                 {},
 		{txnType, "ResetDeadline"}:                   {},
 		{txnType, "Run"}:                             {},
 		{txnType, "Send"}:                            {},
@@ -427,6 +428,8 @@ func TestCommonMethods(t *testing.T) {
 		{txnType, "UserPriority"}:                    {},
 		{txnType, "AnchorKey"}:                       {},
 		{txnType, "ID"}:                              {},
+		{txnType, "IsAborted"}:                       {},
+		{txnType, "IsCommitted"}:                     {},
 	}
 
 	for b := range omittedChecks {

--- a/pkg/internal/client/txn.go
+++ b/pkg/internal/client/txn.go
@@ -154,10 +154,21 @@ func (txn *Txn) IsFinalized() bool {
 	return txn.mu.finalized
 }
 
+// status returns the txn proto status field.
 func (txn *Txn) status() roachpb.TransactionStatus {
 	txn.mu.Lock()
 	defer txn.mu.Unlock()
 	return txn.mu.Proto.Status
+}
+
+// IsCommitted returns true if the transaction has the committed status.
+func (txn *Txn) IsCommitted() bool {
+	return txn.status() == roachpb.COMMITTED
+}
+
+// IsAborted returns true if the transaction has the aborted status.
+func (txn *Txn) IsAborted() bool {
+	return txn.status() == roachpb.ABORTED
 }
 
 // SetUserPriority sets the transaction's user priority. Transactions default to
@@ -709,6 +720,14 @@ func (e *AutoCommitError) Error() string {
 // to clean up the transaction before returning an error. In case of
 // TransactionAbortedError, txn is reset to a fresh transaction, ready to be
 // used.
+//
+// TODO(andrei): The SQL Executor was the most complex user of this interface.
+// It needed fine control by using TxnExecOptions. Now SQL no longer uses this
+// interface, so it's time to see how it can be simplified. TxnExecOptions can
+// probably go away, and so can AutoCommitError. The method should also be
+// documented to not allow calls concurrent with any other txn use, so that the
+// Commit() call inside it is clearly correct (as in, it won't run concurrently
+// with other txn calls).
 func (txn *Txn) Exec(
 	ctx context.Context, opt TxnExecOptions, fn func(context.Context, *Txn, *TxnExecOptions) error,
 ) (err error) {
@@ -722,11 +741,7 @@ func (txn *Txn) Exec(
 		err = fn(ctx, txn, &opt)
 
 		if err == nil && opt.AutoCommit {
-			// Copy the status out of the Proto under lock. Making decisions on
-			// this later is not thread-safe, but the commutativity property of
-			// transactions assure that reasoning about the situation is straightforward.
 			status := txn.status()
-
 			switch status {
 			case roachpb.ABORTED:
 				// TODO(andrei): Until 7881 is fixed.
@@ -766,13 +781,20 @@ func (txn *Txn) Exec(
 			break
 		}
 
-		txn.commitTriggers = nil
-
-		log.VEventf(ctx, 2, "automatically retrying transaction: %s because of error: %s",
-			txn.DebugName(), err)
+		txn.PrepareForRetry(ctx, err)
 	}
 
 	return err
+}
+
+// PrepareForRetry needs to be called before an retry to perform some
+// book-keeping.
+//
+// TODO(andrei): I think this is called in the wrong place. See #18170.
+func (txn *Txn) PrepareForRetry(ctx context.Context, err error) {
+	txn.commitTriggers = nil
+	log.VEventf(ctx, 2, "automatically retrying transaction: %s because of error: %s",
+		txn.DebugName(), err)
 }
 
 // IsRetryableErrMeantForTxn returns true if err is a retryable

--- a/pkg/sql/copy.go
+++ b/pkg/sql/copy.go
@@ -346,4 +346,4 @@ func (CopyDataBlock) StatementType() parser.StatementType { return parser.RowsAf
 
 // StatementTag returns a short string identifying the type of statement.
 func (CopyDataBlock) StatementTag() string { return "" }
-func (CopyDataBlock) String() string       { return "" }
+func (CopyDataBlock) String() string       { return "CopyDataBlock" }

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -344,6 +344,19 @@ type ExecutorTestingKnobs struct {
 
 	// DistSQLPlannerKnobs are testing knobs for distSQLPlanner.
 	DistSQLPlannerKnobs DistSQLPlannerTestingKnobs
+
+	// BeforeAutoCommit is called when the Executor is about to commit the KV
+	// transaction after running a statement in an implicit transaction, allowing
+	// tests to inject errors into that commit.
+	// If an error is returned, that error will be considered the result of
+	// txn.Commit(), and the txn.Commit() call will not actually be
+	// made. If no error is returned, txn.Commit() is called normally.
+	//
+	// Note that this is not called if the SQL statement representing the implicit
+	// transaction has committed the KV txn itself (e.g. if it used the 1-PC
+	// optimization). This is only called when the Executor is the one doing the
+	// committing.
+	BeforeAutoCommit func(ctx context.Context, stmt string) error
 }
 
 // DistSQLPlannerTestingKnobs is used to control internals of the distSQLPlanner
@@ -569,6 +582,9 @@ func (e *Executor) Prepare(
 // Note that we will only receive an error even if we run a successful statement
 // followed by a statement which has an error then the caller will only receive
 // the error, however the first statement will have been executed.
+//
+// If no error is returned, the caller has to call Close() on the returned
+// StatementResults.
 func (e *Executor) ExecuteStatementsBuffered(
 	session *Session, stmts string, pinfo *parser.PlaceholderInfo, expectedNumResults int,
 ) (StatementResults, error) {
@@ -741,7 +757,7 @@ func (e *Executor) execRequest(
 		// statements or only one, so just pretend there was one.
 		if txnState.mu.txn != nil {
 			// Rollback the txn.
-			txnState.updateStateAndCleanupOnErr(err, e)
+			err = txnState.updateStateAndCleanupOnErr(err, e)
 		}
 		return err
 	}
@@ -767,7 +783,6 @@ func (e *Executor) execParsed(
 		// that is encountered resets stmts.
 
 		inTxn := txnState.State() != NoTxn
-		execOpt := client.TxnExecOptions{}
 		// Figure out the statements out of which we're going to try to consume
 		// this iteration. If we need to create an implicit txn, only one statement
 		// can be consumed.
@@ -775,13 +790,18 @@ func (e *Executor) execParsed(
 		// If protoTS is set, the transaction proto sets its Orig and Max timestamps
 		// to it each retry.
 		var protoTS *hlc.Timestamp
+
+		// autoCommit will be set if we're now starting an implicit transaction and
+		// thus should automatically commit after we've run the statement.
+		autoCommit := false
+
 		// If we're not in a transaction, then the next statement is part of a new
 		// transaction (implicit txn or explicit txn). We do the corresponding state
 		// reset.
 		if !inTxn {
 			// Detect implicit transactions - they need to be autocommitted.
 			if _, isBegin := stmts[0].AST.(*parser.BeginTransaction); !isBegin {
-				execOpt.AutoCommit = true
+				autoCommit = true
 				stmtsToExec = stmtsToExec[:1]
 				// Check for AS OF SYSTEM TIME. If it is present but not detected here,
 				// it will raise an error later on.
@@ -800,14 +820,13 @@ func (e *Executor) execParsed(
 			}
 			txnState.resetForNewSQLTxn(
 				e, session,
-				execOpt.AutoCommit, /* implicitTxn */
-				false,              /* retryIntent */
+				autoCommit, /* implicitTxn */
+				false,      /* retryIntent */
 				e.cfg.Clock.PhysicalTime(), /* sqlTimestamp */
 				session.DefaultIsolationLevel,
 				roachpb.NormalUserPriority,
 			)
 		}
-		execOpt.AutoRetry = txnState.State() == AutoRetry
 
 		if txnState.State() == NoTxn {
 			panic("we failed to initialize a txn")
@@ -818,15 +837,9 @@ func (e *Executor) execParsed(
 
 		// Track if we are retrying this query, so that we do not double count.
 		automaticRetryCount := 0
-		txnClosure := func(ctx context.Context, txn *client.Txn, opt *client.TxnExecOptions) error {
+
+		txnClosure := func(ctx context.Context) (transitionToOpen bool, _ error) {
 			defer func() { automaticRetryCount++ }()
-			if txnState.TxnIsOpen() && txnState.mu.txn != txn {
-				panic(fmt.Sprintf("closure wasn't called in the txn we set up for it."+
-					"\ntxnState.mu.txn:%+v\ntxn:%+v\ntxnState:%+v", txnState.mu.txn, txn, txnState))
-			}
-			txnState.mu.Lock()
-			txnState.mu.txn = txn
-			txnState.mu.Unlock()
 
 			if protoTS != nil {
 				txnState.mu.txn.SetFixedTimestamp(*protoTS)
@@ -835,7 +848,7 @@ func (e *Executor) execParsed(
 			// Some results may have been produced by a previous attempt.
 			txnState.txnResults.Reset(session.Ctx())
 			var err error
-			remainingStmts, err = runTxnAttempt(
+			remainingStmts, transitionToOpen, err = runTxnAttempt(
 				e, session, stmtsToExec, pinfo, origState,
 				!inTxn /* txnPrefix */, avoidCachedDescriptors, automaticRetryCount, txnState.txnResults)
 
@@ -854,42 +867,129 @@ func (e *Executor) execParsed(
 				}
 			}
 
-			return err
+			return transitionToOpen, err
 		}
-		// This is where the magic happens - we ask db to run a KV txn and possibly retry it.
-		txn := txnState.mu.txn // this might be nil if the txn was already aborted.
-		err := txn.Exec(session.Ctx(), execOpt, txnClosure)
+
+		// We now run the prefix of stmtsToExec that belong to the current
+		// transaction via txnClosure. We do this in a loop because we may want to
+		// do automatic retries, if possible. Whether or not we can do auto-retries
+		// depends on the state before we execute statements in the batch.
+		//
+		// TODO(andrei): It's unfortunate that we're keeping track of what the state
+		// was before running the statements. It'd be great if the state in which we
+		// find ourselves after running the statements told us if we can auto-retry.
+		// A way to do that is to introduce another state called AutoRestartWait,
+		// similar to RestartWait. We'd enter this state whenever we're in state
+		// AutoRetry and we get a retriable error.
+		txnCanBeAutoRetried := txnState.State() == AutoRetry
+		var err error
+		for {
+			// transitionToOpen will tell us if we need to transition from AutoRetry
+			// to Open after executing the statements.
+			var transitionToOpen bool
+
+			// Run some statements.
+			transitionToOpen, err = txnClosure(session.Ctx())
+
+			// Sanity checks.
+			if err != nil && txnState.TxnIsOpen() {
+				log.Fatalf(session.Ctx(), "statement(s) generated error but state is %s. err: %s",
+					txnState.State(), err)
+			} else {
+				state := txnState.State()
+				if err == nil && (state == Aborted || state == RestartWait) {
+					crash := true
+					if len(stmtsToExec) > 0 {
+						// SHOW TRANSACTION STATUS statements are always allowed regardless
+						// of the transaction state.
+						if _, ok := stmtsToExec[0].AST.(*parser.ShowTransactionStatus); ok {
+							crash = false
+						}
+					}
+					if crash {
+						log.Fatalf(session.Ctx(), "no error, but we're in error state: %s", state)
+					}
+				}
+			}
+
+			// Check if we need to auto-commit. If so, we end the transaction now; the
+			// transaction was only supposed to exist for the statement that we just
+			// ran.
+			if autoCommit {
+				if err == nil {
+					txn := txnState.mu.txn
+					if txn == nil {
+						log.Fatalf(session.Ctx(), "implicit txn returned with no error and yet "+
+							"the kv txn is gone. No state transition should have done that. State: %s",
+							txnState.State())
+					}
+
+					// We're were told to autoCommit. The KV txn might already be
+					// committed (planNodes are free to do that when running an implicit
+					// transaction, and some try to do it to take advantage of 1-PC txns).
+					// If it is, then there's nothing to do. If it isn't, then we commit
+					// it here.
+					//
+					// NOTE(andrei): It bothers me some that we're peeking at txn to
+					// figure out whether we committed or not, where SQL could already
+					// know that - individual statements could report this back.
+					if txn.IsAborted() {
+						// TODO(andrei): Until 7881 is fixed.
+						log.Fatalf(session.Ctx(), "#7881: the statement we just ran didn't generate an error "+
+							"but the txn proto is aborted. This should never happen. txn: %+v",
+							txn)
+					}
+
+					if !txn.IsCommitted() {
+						var skipCommit bool
+						if e.cfg.TestingKnobs.BeforeAutoCommit != nil {
+							err = e.cfg.TestingKnobs.BeforeAutoCommit(session.Ctx(), stmtsToExec[0].String())
+							skipCommit = err != nil
+						}
+						if !skipCommit {
+							err = txn.Commit(session.Ctx())
+						}
+						log.Eventf(session.Ctx(), "AutoCommit. err: %v\ntxn: %+v", err, txn.Proto())
+						if err != nil {
+							err = txnState.updateStateAndCleanupOnErr(err, e)
+						}
+					}
+				}
+
+				// After autoCommit, unless we're in RestartWait, we leave the
+				// transaction in NoTxn, regardless of whether we executed the query
+				// successfully or we encountered an error.
+				if txnState.State() != RestartWait {
+					txnState.resetStateAndTxn(NoTxn)
+				}
+			}
+
+			// If we've been told by runTxnAttempt that we should move to Open, do it
+			// now.
+			if err == nil && (txnState.State() == AutoRetry) && transitionToOpen {
+				txnState.SetState(Open)
+			}
+
+			if _, ok := err.(*roachpb.UnhandledRetryableError); ok {
+				// We sent transactional requests, so the TxnCoordSender was supposed to
+				// turn retryable errors into HandledRetryableTxnError.
+				log.Fatalf(session.Ctx(), "unexpected UnhandledRetryableError at the Executor level: %s", err)
+			}
+
+			shouldAutoRetry := txnState.State() == RestartWait && txnCanBeAutoRetried
+			if !shouldAutoRetry {
+				break
+			}
+			txnState.mu.txn.PrepareForRetry(session.Ctx(), err)
+		}
+		// We're now done with auto-retries.
+
 		if err != nil && (log.V(2) || logStatementsExecuteEnabled.Get(&e.cfg.Settings.SV)) {
 			log.Infof(session.Ctx(), "execParsed: error: %v. state: %s", err, txnState.State())
 		}
 
-		// Update the Err field of the last result if the error was coming from
-		// auto commit. The error was generated outside of the txn closure, so it was not
-		// set in any result.
-		if err != nil {
-			if aErr, ok := err.(*client.AutoCommitError); ok {
-				// TODO(andrei): Until #7881 fixed.
-				{
-					if txnState.mu.txn != nil {
-						log.Eventf(session.Ctx(), "executor got AutoCommitError: %s\n"+
-							"txn: %+v\nexecOpt.AutoRetry %t, execOpt.AutoCommit:%t, stmts %+v, remaining %+v",
-							aErr, txnState.mu.txn.Proto(), execOpt.AutoRetry, execOpt.AutoCommit, stmts,
-							remainingStmts)
-					} else {
-						log.Errorf(session.Ctx(), "7881: AutoCommitError on nil txn: %s, "+
-							"txnState %+v, execOpt %+v, stmts %+v, remaining %+v, txn captured before executing batch: %+v",
-							aErr, txnState, execOpt, stmts, remainingStmts, txn)
-						txnState.sp.SetBaggageItem(keyFor7881Sample, "sample me please")
-					}
-				}
-				e.TxnAbortCount.Inc(1)
-				// TODO(andrei): Once 7881 is fixed, this should be
-				// txnState.mu.txn.CleanupOnError().
-				txn.CleanupOnError(session.Ctx(), err)
-			}
-		}
-
-		// Sanity check about not leaving KV txns open on errors.
+		// Sanity check about not leaving KV txns open on errors (other than
+		// retriable errors).
 		if err != nil && txnState.mu.txn != nil && !txnState.mu.txn.IsFinalized() {
 			if _, retryable := err.(*roachpb.HandledRetryableTxnError); !retryable {
 				log.Fatalf(session.Ctx(), "got a non-retryable error but the KV "+
@@ -898,23 +998,15 @@ func (e *Executor) execParsed(
 			}
 		}
 
-		// Now make sense of the state we got into and update txnState.
-		if (txnState.State() == RestartWait || txnState.State() == Aborted) &&
-			txnState.commitSeen {
-			// A COMMIT got an error (retryable or not). Too bad, this txn is toast.
+		if txnState.commitSeen && txnState.State() == Aborted {
+			// A COMMIT got an error (retryable or not); we'll move to state NoTxn.
 			// After we return a result for COMMIT (with the COMMIT pgwire tag), the
 			// user can't send any more commands.
-			e.TxnAbortCount.Inc(1)
-			txn.CleanupOnError(session.Ctx(), err)
 			txnState.resetStateAndTxn(NoTxn)
 		}
 
-		if execOpt.AutoCommit {
-			// If execOpt.AutoCommit was set, then the txn no longer exists at this point.
-			txnState.resetStateAndTxn(NoTxn)
-		}
-
-		// If we're no longer in a transaction, finish the trace.
+		// If we're no longer in a transaction, close the transaction-scoped
+		// resources.
 		if txnState.State() == NoTxn {
 			txnState.finishSQLTxn(session)
 		}
@@ -971,7 +1063,7 @@ func (e *Executor) execParsed(
 		// Figure out what statements to run on the next iteration.
 		if err != nil {
 			return convertToErrWithPGCode(err)
-		} else if execOpt.AutoCommit {
+		} else if autoCommit {
 			stmts = stmts[1:]
 		} else {
 			stmts = remainingStmts
@@ -981,25 +1073,27 @@ func (e *Executor) execParsed(
 	return nil
 }
 
-// runTxnAttempt is used in the closure we pass to txn.Exec(). It will be called
-// possibly multiple times with the same set to statements.
-// It takes in a batch of statements and executes a prefix of them corresponding
-// to one transaction. More technically, it runs them one by one until either:
+// runTxnAttempt takes in a batch of statements and executes a prefix of them
+// corresponding to one transaction. It is called multiple times with the same
+// set of statements when performing auto-retries.
+//
+// More technically, it run statements one by one until either:
 // - a statement returns an error
 // - the end of the the batch is reached
 // - a statement transition the state to a non-"open" one, meaning that we're
 // not in a transaction any more (or we are in a transaction but in an error
 // state, although this also triggers the error condition above).
 //
-// This method may transition the state from AutoRetry to Open if the end of the
-// batch is reached.
-//
 // Args:
 // txnPrefix: set if the start of the batch corresponds to the start of the
 // current transaction. Used to trap nested BEGINs.
 // txnResults: used to push query results.
 //
-// It returns all the statements that were not executed.
+// It returns:
+// remainingStmts: all the statements that were not executed.
+// transitionToOpen: specifies if the caller should move from state AutoRetry to
+// state Open. This will be false if the state is not AutoRetry when this
+// returns.
 func runTxnAttempt(
 	e *Executor,
 	session *Session,
@@ -1010,7 +1104,7 @@ func runTxnAttempt(
 	avoidCachedDescriptors bool,
 	automaticRetryCount int,
 	txnResults ResultsGroup,
-) (StatementList, error) {
+) (remainingStmts StatementList, transitionToOpen bool, _ error) {
 
 	txnState := &session.TxnState
 
@@ -1048,7 +1142,7 @@ func runTxnAttempt(
 			txnPrefix && i == 0, /* firstInTxn */
 			avoidCachedDescriptors, automaticRetryCount, stmtResult,
 		); err != nil {
-			return nil, err
+			return nil, false, err
 		}
 
 		// If the transaction is done, stop executing more statements.
@@ -1062,20 +1156,19 @@ func runTxnAttempt(
 	// statements. We also want future ResultsSentToClient() calls to return
 	// false.
 	if err := txnState.txnResults.Flush(session.Ctx()); err != nil {
-		txnState.updateStateAndCleanupOnErr(err, e)
-		return nil, err
+		err = txnState.updateStateAndCleanupOnErr(err, e)
+		return nil, false, err
 	}
 
 	// If we're still in AutoRetry but we've seen statements that would need to be
 	// retried if we'd later want to retry the transaction, we transition to Open.
 	// Automatic retries are not possible beyond this point.
-	if txnState.State() == AutoRetry && encounteredNonAutoRetryStmt {
-		txnState.SetState(Open)
-	}
+	transitionToOpen = txnState.State() == AutoRetry && encounteredNonAutoRetryStmt
+
 	// Return the statements that weren't executed in the current txn, so they can
 	// be executed in the context of other transactions. The set can be empty if
 	// we executed everything that was passed in.
-	return stmts[i+1:], nil
+	return stmts[i+1:], transitionToOpen, nil
 }
 
 // execSingleStatement executes one statement by dispatching according to the
@@ -1277,14 +1370,14 @@ func (e *Executor) execStmtInAbortedTxn(
 		}
 		if err := parser.ValidateRestartCheckpoint(spName); err != nil {
 			if txnState.State() == RestartWait {
-				txnState.updateStateAndCleanupOnErr(err, e)
+				err = txnState.updateStateAndCleanupOnErr(err, e)
 			}
 			return err
 		}
 		if !txnState.retryIntent {
 			err := fmt.Errorf("SAVEPOINT %s has not been used", parser.RestartSavepointName)
 			if txnState.State() == RestartWait {
-				txnState.updateStateAndCleanupOnErr(err, e)
+				err = txnState.updateStateAndCleanupOnErr(err, e)
 			}
 			return err
 		}
@@ -1322,8 +1415,7 @@ func (e *Executor) execStmtInAbortedTxn(
 			// there seems to be little point in staying in RestartWait (plus,
 			// higher-level code asserts that we're only in RestartWait when returning
 			// retryable errors to the client).
-			txnState.updateStateAndCleanupOnErr(err, e)
-			return err
+			return txnState.updateStateAndCleanupOnErr(err, e)
 		}
 		return sqlbase.NewTransactionAbortedError("" /* customMsg */)
 	}
@@ -1411,7 +1503,7 @@ func (e *Executor) execStmtInOpenTxn(
 			if !txnState.TxnIsOpen() {
 				panic(fmt.Sprintf("unexpected txnState when cleaning up: %v", txnState.State()))
 			}
-			txnState.updateStateAndCleanupOnErr(err, e)
+			err = txnState.updateStateAndCleanupOnErr(err, e)
 
 			if firstInTxn && isBegin(stmt) {
 				// A failed BEGIN statement that was starting a txn doesn't leave the
@@ -1455,6 +1547,7 @@ func (e *Executor) execStmtInOpenTxn(
 				// client.Txn.
 				roachpb.Transaction{},
 			)
+			err = txnState.updateStateAndCleanupOnErr(err, e)
 		}
 	}()
 

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -831,158 +831,23 @@ func (e *Executor) execParsed(
 		if txnState.State() == NoTxn {
 			panic("we failed to initialize a txn")
 		}
-		// Now actually run some statements.
-		var remainingStmts StatementList
-		origState := txnState.State()
 
-		// Track if we are retrying this query, so that we do not double count.
-		automaticRetryCount := 0
-
-		txnClosure := func(ctx context.Context) (transitionToOpen bool, _ error) {
-			defer func() { automaticRetryCount++ }()
-
-			if protoTS != nil {
-				txnState.mu.txn.SetFixedTimestamp(*protoTS)
-			}
-
-			// Some results may have been produced by a previous attempt.
-			txnState.txnResults.Reset(session.Ctx())
-			var err error
-			remainingStmts, transitionToOpen, err = runTxnAttempt(
-				e, session, stmtsToExec, pinfo, origState,
-				!inTxn /* txnPrefix */, avoidCachedDescriptors, automaticRetryCount, txnState.txnResults)
-
-			// TODO(andrei): Until #7881 fixed.
-			if err == nil && txnState.State() == Aborted {
-				doWarn := true
-				if len(stmtsToExec) > 0 {
-					if _, ok := stmtsToExec[0].AST.(*parser.ShowTransactionStatus); ok {
-						doWarn = false
-					}
-				}
-				if doWarn {
-					log.Errorf(ctx,
-						"7881: txnState is Aborted without an error propagating. stmtsToExec: %s, "+
-							"remainingStmts: %s, txnState: %+v", stmtsToExec, remainingStmts, txnState)
-				}
-			}
-
-			return transitionToOpen, err
-		}
-
-		// We now run the prefix of stmtsToExec that belong to the current
-		// transaction via txnClosure. We do this in a loop because we may want to
-		// do automatic retries, if possible. Whether or not we can do auto-retries
-		// depends on the state before we execute statements in the batch.
-		//
-		// TODO(andrei): It's unfortunate that we're keeping track of what the state
-		// was before running the statements. It'd be great if the state in which we
-		// find ourselves after running the statements told us if we can auto-retry.
-		// A way to do that is to introduce another state called AutoRestartWait,
-		// similar to RestartWait. We'd enter this state whenever we're in state
-		// AutoRetry and we get a retriable error.
-		txnCanBeAutoRetried := txnState.State() == AutoRetry
 		var err error
-		for {
-			// transitionToOpen will tell us if we need to transition from AutoRetry
-			// to Open after executing the statements.
-			var transitionToOpen bool
-
-			// Run some statements.
-			transitionToOpen, err = txnClosure(session.Ctx())
-
-			// Sanity checks.
-			if err != nil && txnState.TxnIsOpen() {
-				log.Fatalf(session.Ctx(), "statement(s) generated error but state is %s. err: %s",
-					txnState.State(), err)
-			} else {
-				state := txnState.State()
-				if err == nil && (state == Aborted || state == RestartWait) {
-					crash := true
-					if len(stmtsToExec) > 0 {
-						// SHOW TRANSACTION STATUS statements are always allowed regardless
-						// of the transaction state.
-						if _, ok := stmtsToExec[0].AST.(*parser.ShowTransactionStatus); ok {
-							crash = false
-						}
-					}
-					if crash {
-						log.Fatalf(session.Ctx(), "no error, but we're in error state: %s", state)
-					}
-				}
-			}
-
-			// Check if we need to auto-commit. If so, we end the transaction now; the
-			// transaction was only supposed to exist for the statement that we just
-			// ran.
-			if autoCommit {
-				if err == nil {
-					txn := txnState.mu.txn
-					if txn == nil {
-						log.Fatalf(session.Ctx(), "implicit txn returned with no error and yet "+
-							"the kv txn is gone. No state transition should have done that. State: %s",
-							txnState.State())
-					}
-
-					// We're were told to autoCommit. The KV txn might already be
-					// committed (planNodes are free to do that when running an implicit
-					// transaction, and some try to do it to take advantage of 1-PC txns).
-					// If it is, then there's nothing to do. If it isn't, then we commit
-					// it here.
-					//
-					// NOTE(andrei): It bothers me some that we're peeking at txn to
-					// figure out whether we committed or not, where SQL could already
-					// know that - individual statements could report this back.
-					if txn.IsAborted() {
-						// TODO(andrei): Until 7881 is fixed.
-						log.Fatalf(session.Ctx(), "#7881: the statement we just ran didn't generate an error "+
-							"but the txn proto is aborted. This should never happen. txn: %+v",
-							txn)
-					}
-
-					if !txn.IsCommitted() {
-						var skipCommit bool
-						if e.cfg.TestingKnobs.BeforeAutoCommit != nil {
-							err = e.cfg.TestingKnobs.BeforeAutoCommit(session.Ctx(), stmtsToExec[0].String())
-							skipCommit = err != nil
-						}
-						if !skipCommit {
-							err = txn.Commit(session.Ctx())
-						}
-						log.Eventf(session.Ctx(), "AutoCommit. err: %v\ntxn: %+v", err, txn.Proto())
-						if err != nil {
-							err = txnState.updateStateAndCleanupOnErr(err, e)
-						}
-					}
-				}
-
-				// After autoCommit, unless we're in RestartWait, we leave the
-				// transaction in NoTxn, regardless of whether we executed the query
-				// successfully or we encountered an error.
-				if txnState.State() != RestartWait {
-					txnState.resetStateAndTxn(NoTxn)
-				}
-			}
-
-			// If we've been told by runTxnAttempt that we should move to Open, do it
-			// now.
-			if err == nil && (txnState.State() == AutoRetry) && transitionToOpen {
-				txnState.SetState(Open)
-			}
-
-			if _, ok := err.(*roachpb.UnhandledRetryableError); ok {
-				// We sent transactional requests, so the TxnCoordSender was supposed to
-				// turn retryable errors into HandledRetryableTxnError.
-				log.Fatalf(session.Ctx(), "unexpected UnhandledRetryableError at the Executor level: %s", err)
-			}
-
-			shouldAutoRetry := txnState.State() == RestartWait && txnCanBeAutoRetried
-			if !shouldAutoRetry {
-				break
-			}
-			txnState.mu.txn.PrepareForRetry(session.Ctx(), err)
+		var remainingStmts StatementList
+		var transitionToOpen bool
+		remainingStmts, transitionToOpen, err = runWithAutoRetry(
+			e, session, stmtsToExec, !inTxn /* txnPrefix */, autoCommit,
+			protoTS, pinfo, avoidCachedDescriptors,
+		)
+		if autoCommit && txnState.State() != NoTxn {
+			log.Fatalf(session.Ctx(), "after an implicit txn, state should always be NoTxn, but found: %s",
+				txnState.State())
 		}
-		// We're now done with auto-retries.
+
+		// If we've been told that we should move to Open, do it now.
+		if err == nil && (txnState.State() == AutoRetry) && transitionToOpen {
+			txnState.SetState(Open)
+		}
 
 		if err != nil && (log.V(2) || logStatementsExecuteEnabled.Get(&e.cfg.Settings.SV)) {
 			log.Infof(session.Ctx(), "execParsed: error: %v. state: %s", err, txnState.State())
@@ -1071,6 +936,184 @@ func (e *Executor) execParsed(
 	}
 
 	return nil
+}
+
+// runWithAutoRetry runs a prefix of stmtsToExec corresponding to the current
+// transaction. It deals with auto-retries: when possible, the statements are
+// retried in case of retriable errors. It also deals with "autoCommit" - if the
+// current transaction only consists of a single statement (i.e. it's an
+// "implicit txn"), then this function deal with committing the transaction and
+// possibly retrying it if the commit gets a retriable error.
+//
+// Args:
+// stmtsToExec: A prefix of these will be executed. The remaining ones will be
+// returned as remainingStmts.
+// txnPrefix: Set if stmtsToExec corresponds to the start of the current
+// transaction. Used to trap nested BEGINs.
+// autoCommit: Set if the transction should be committed at the end of the
+// protoTS: If not nil, the transaction proto sets its Orig and Max timestamps
+// to it each retry.
+//
+// Returns:
+// remainingStmts: all the statements that were not executed.
+// transitionToOpen: specifies if the caller should move from state AutoRetry to
+// state Open. This will be false if the state is not AutoRetry when this
+// returns.
+// err: An error that occurred while executing the queries.
+func runWithAutoRetry(
+	e *Executor,
+	session *Session,
+	stmtsToExec StatementList,
+	txnPrefix bool,
+	autoCommit bool,
+	protoTS *hlc.Timestamp,
+	pinfo *parser.PlaceholderInfo,
+	avoidCachedDescriptors bool,
+) (remainingStmts StatementList, transitionToOpen bool, _ error) {
+
+	if autoCommit && !txnPrefix {
+		log.Fatal(session.Ctx(), "autoCommit implies txnPrefix. "+
+			"How could the transaction have been started before an implicit txn?")
+	}
+	if autoCommit && len(stmtsToExec) != 1 {
+		log.Fatal(session.Ctx(), "Only one statement should be executed when "+
+			"autoCommit is set. stmtsToExec: %s", stmtsToExec)
+	}
+
+	txnState := &session.TxnState
+	origState := txnState.State()
+
+	// Whether or not we can do auto-retries depends on the state before we
+	// execute statements in the batch.
+	//
+	// TODO(andrei): It's unfortunate that we're keeping track of what the state
+	// was before running the statements. It'd be great if the state in which we
+	// find ourselves after running the statements told us if we can auto-retry.
+	// A way to do that is to introduce another state called AutoRestartWait,
+	// similar to RestartWait. We'd enter this state whenever we're in state
+	// AutoRetry and we get a retriable error.
+	txnCanBeAutoRetried := txnState.State() == AutoRetry
+	if autoCommit && !txnCanBeAutoRetried {
+		log.Fatalf(session.Ctx(), "autoCommit is only supported in the AutoRetry state. "+
+			"Current state: %s", txnState.State())
+	}
+
+	// Track if we are retrying this query, so that we do not double count.
+	automaticRetryCount := 0
+
+	var err error
+	for {
+
+		if protoTS != nil {
+			txnState.mu.txn.SetFixedTimestamp(*protoTS)
+		}
+		// Some results may have been produced by a previous attempt.
+		txnState.txnResults.Reset(session.Ctx())
+
+		// Run some statements.
+		remainingStmts, transitionToOpen, err = runTxnAttempt(
+			e, session, stmtsToExec, pinfo, origState,
+			txnPrefix, avoidCachedDescriptors, automaticRetryCount, txnState.txnResults)
+
+		// Sanity checks.
+		if err != nil && txnState.TxnIsOpen() {
+			log.Fatalf(session.Ctx(), "statement(s) generated error but state is %s. err: %s",
+				txnState.State(), err)
+		} else {
+			state := txnState.State()
+			if err == nil && (state == Aborted || state == RestartWait) {
+				crash := true
+				// SHOW TRANSACTION STATUS statements are always allowed regardless of
+				// the transaction state.
+				if len(stmtsToExec) > 0 {
+					if _, ok := stmtsToExec[0].AST.(*parser.ShowTransactionStatus); ok {
+						crash = false
+					}
+				}
+				if crash {
+					log.Fatalf(session.Ctx(), "no error, but we're in error state: %s", state)
+				}
+			}
+		}
+
+		// Check if we need to auto-commit. If so, we end the transaction now; the
+		// transaction was only supposed to exist for the statement that we just
+		// ran.
+		if autoCommit {
+			if err == nil {
+				txn := txnState.mu.txn
+				if txn == nil {
+					log.Fatalf(session.Ctx(), "implicit txn returned with no error and yet "+
+						"the kv txn is gone. No state transition should have done that. State: %s",
+						txnState.State())
+				}
+
+				// We were told to autoCommit. The KV txn might already be committed
+				// (planNodes are free to do that when running an implicit transaction,
+				// and some try to do it to take advantage of 1-PC txns). If it is, then
+				// there's nothing to do. If it isn't, then we commit it here.
+				//
+				// NOTE(andrei): It bothers me some that we're peeking at txn to figure
+				// out whether we committed or not, where SQL could already know that -
+				// individual statements could report this back.
+				if txn.IsAborted() {
+					log.Fatalf(session.Ctx(), "#7881: the statement we just ran didn't generate an error "+
+						"but the txn proto is aborted. This should never happen. txn: %+v",
+						txn)
+				}
+
+				if !txn.IsCommitted() {
+					var skipCommit bool
+					if e.cfg.TestingKnobs.BeforeAutoCommit != nil {
+						err = e.cfg.TestingKnobs.BeforeAutoCommit(session.Ctx(), stmtsToExec[0].String())
+						skipCommit = err != nil
+					}
+					if !skipCommit {
+						err = txn.Commit(session.Ctx())
+					}
+					log.Eventf(session.Ctx(), "AutoCommit. err: %v\ntxn: %+v", err, txn.Proto())
+					if err != nil {
+						err = txnState.updateStateAndCleanupOnErr(err, e)
+					}
+				}
+			}
+
+			// After autoCommit, unless we're in RestartWait, we leave the transaction
+			// in NoTxn, regardless of whether we executed the query successfully or
+			// we encountered an error.
+			if txnState.State() != RestartWait {
+				txnState.resetStateAndTxn(NoTxn)
+			}
+		}
+
+		if _, ok := err.(*roachpb.UnhandledRetryableError); ok {
+			// We sent transactional requests, so the TxnCoordSender was supposed to
+			// turn retryable errors into HandledRetryableTxnError.
+			log.Fatalf(session.Ctx(), "unexpected UnhandledRetryableError at the Executor level: %s", err)
+		}
+
+		resultsSentToClient := txnState.txnResults.ResultsSentToClient()
+		shouldAutoRetry := txnState.State() == RestartWait && txnCanBeAutoRetried
+		if shouldAutoRetry && resultsSentToClient {
+			shouldAutoRetry = false
+			// We otherwise can and should auto-retry, but, alas, we've already sent
+			// some results to the client, so we can no longer auto-retry. We only
+			// stay in RestartWait if the client is doing client-directed retries.
+			// Otherwise, we move to Aborted.
+			if !txnState.retryIntent {
+				e.TxnAbortCount.Inc(1)
+				txnState.mu.txn.CleanupOnError(session.Ctx(), err)
+				txnState.resetStateAndTxn(Aborted)
+			}
+		}
+
+		if !shouldAutoRetry {
+			break
+		}
+		txnState.mu.txn.PrepareForRetry(session.Ctx(), err)
+		automaticRetryCount++
+	}
+	return remainingStmts, transitionToOpen, err
 }
 
 // runTxnAttempt takes in a batch of statements and executes a prefix of them

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -1210,7 +1210,12 @@ func (e *Executor) execSingleStatement(
 // getTransactionState retrieves a text representation of the given state.
 func getTransactionState(txnState *txnState) string {
 	state := txnState.State()
-	if txnState.implicitTxn {
+	// If the statement reading the state is in an implicit transaction, then we
+	// want to tell NoTxn to the client.
+	// If implicitTxn is set, the state is supposed to be AutoRetry. However, we
+	// test the state too, just in case we have a state machine bug, in which case
+	// we don't want this code to hide a bug.
+	if txnState.implicitTxn && state == AutoRetry {
 		state = NoTxn
 	}
 	// For the purposes of representing the states to client, make the AutoRetry

--- a/pkg/sql/session.go
+++ b/pkg/sql/session.go
@@ -1186,9 +1186,16 @@ func (ts *txnState) finishSQLTxn(s *Session) {
 }
 
 // updateStateAndCleanupOnErr updates txnState based on the type of error that we
-// received. If it's a retriable error and we're going to retry the txn,
-// then the state moves to RestartWait. Otherwise, the state moves to Aborted
-// and the KV txn is cleaned up.
+// received. If it's a retriable error and it looks like we're going to retry
+// the txn (we're either in the AutoRetry state, meaning that we can do
+// auto-retries, or the client is doing client-directed retries), then the state
+// moves to RestartWait. Otherwise, the state moves to Aborted and the KV txn is
+// cleaned up.
+// Note that even if we move to RestartWait here, this doesn't automatically
+// mean that we're going to auto-retry. It might be the case, for example, that
+// we've already streamed results to the client and so we can't auto-retry for
+// that reason. It is the responsibility of higher layers to catch this and
+// terminate the transaction, if appropriate.
 //
 // This method always returns an error. Usually it's the input err, except that
 // a retriable error meant for another txn is replaced with a non-retriable

--- a/pkg/sql/session.go
+++ b/pkg/sql/session.go
@@ -530,13 +530,13 @@ func (s *Session) Finish(e *Executor) {
 	// TODO(nvanbenschoten): Once we have better support for cancelling ongoing
 	// statement execution by the infrastructure added to support CancelRequest,
 	// we should try to actively drain this queue instead of passively waiting
-	// for it to drain.
+	// for it to drain. (andrei, 2017/09) - We now have support for statement
+	// cancellation. Now what?
 	_ = s.synchronizeParallelStmts(s.context)
 
 	// If we're inside a txn, roll it back.
 	if s.TxnState.State().kvTxnIsOpen() {
-		s.TxnState.updateStateAndCleanupOnErr(
-			errors.Errorf("session closing"), e)
+		_ = s.TxnState.updateStateAndCleanupOnErr(fmt.Errorf("session closing"), e)
 	}
 	if s.TxnState.State() != NoTxn {
 		s.TxnState.finishSQLTxn(s)
@@ -1135,7 +1135,7 @@ func (ts *txnState) resetStateAndTxn(state TxnStateEnum) {
 	}
 	if ts.mu.txn != nil && !ts.mu.txn.IsFinalized() {
 		panic(fmt.Sprintf(
-			"attempting to move SQL txn to state %v inconsistent with KV txn state: %s "+
+			"attempting to move SQL txn to state %s inconsistent with KV txn state: %s "+
 				"(finalized: false)", state, ts.mu.txn.Proto().Status))
 	}
 	ts.SetState(state)
@@ -1189,26 +1189,55 @@ func (ts *txnState) finishSQLTxn(s *Session) {
 // received. If it's a retriable error and we're going to retry the txn,
 // then the state moves to RestartWait. Otherwise, the state moves to Aborted
 // and the KV txn is cleaned up.
-func (ts *txnState) updateStateAndCleanupOnErr(err error, e *Executor) {
+//
+// This method always returns an error. Usually it's the input err, except that
+// a retriable error meant for another txn is replaced with a non-retriable
+// error because higher layers are not supposed to consider it retriable.
+func (ts *txnState) updateStateAndCleanupOnErr(err error, e *Executor) error {
 	if err == nil {
 		panic("updateStateAndCleanupOnErr called with no error")
 	}
+	if ts.mu.txn == nil {
+		panic(fmt.Sprintf(
+			"updateStateAndCleanupOnErr called in state with no KV txn. State: %s",
+			ts.State()))
+	}
 	if retErr, ok := err.(*roachpb.HandledRetryableTxnError); !ok ||
 		!ts.willBeRetried() ||
-		!ts.mu.txn.IsRetryableErrMeantForTxn(*retErr) {
+		!ts.mu.txn.IsRetryableErrMeantForTxn(*retErr) ||
+		// If we ran a COMMIT, then we can only do auto-retries, not client-directed
+		// retries.
+		(ts.commitSeen && ts.State() != AutoRetry) {
 
 		// We can't or don't want to retry this txn, so the txn is over.
 		e.TxnAbortCount.Inc(1)
+
+		// If we got a retriable error but it was meant for another txn, we'll
+		// return a non-retriable error instead further down. We need to identify
+		// this here, before the call to ts.resetStateAndTxn().
+		var retriableErrForAnotherTxn bool
+		txnID := ts.mu.txn.Proto().ID
+		if ok && !ts.mu.txn.IsRetryableErrMeantForTxn(*retErr) {
+			retriableErrForAnotherTxn = true
+		}
+
 		// This call rolls back a PENDING transaction and cleans up all its
 		// intents.
 		ts.mu.txn.CleanupOnError(ts.Ctx, err)
 		ts.resetStateAndTxn(Aborted)
+
+		if retriableErrForAnotherTxn {
+			return errors.Wrapf(
+				retErr,
+				"retryable error from another txn. Current txn ID: %v", txnID)
+		}
 	} else {
 		// If we got a retriable error, move the SQL txn to the RestartWait state.
 		// Note that TransactionAborted is also a retriable error, handled here;
 		// in this case cleanup for the txn has been done for us under the hood.
 		ts.SetState(RestartWait)
 	}
+	return err
 }
 
 func (ts *txnState) setIsolationLevel(isolation enginepb.IsolationType) error {

--- a/pkg/sql/session_test.go
+++ b/pkg/sql/session_test.go
@@ -16,13 +16,16 @@ package sql_test
 
 import (
 	"database/sql/driver"
+	"fmt"
 	"net/url"
+	"strings"
 	"testing"
 	"time"
 
 	"github.com/lib/pq"
 	"golang.org/x/net/context"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
@@ -174,5 +177,47 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
 			}
 
 		})
+	}
+}
+
+// Test two things about non-retriable errors happening when the Executor does
+// an "autoCommit" (i.e. commits the KV txn after running an implicit
+// transaction):
+// 1) The error is reported to the client.
+// 2) The error doesn't leave the session in the Aborted state. After running
+// implicit transactions, the state should always be NoTxn, regardless of any
+// errors.
+func TestNonRetriableErrorOnAutoCommit(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	query := "SELECT 42"
+
+	params := base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			SQLExecutor: &sql.ExecutorTestingKnobs{
+				BeforeAutoCommit: func(ctx context.Context, stmt string) error {
+					if strings.Contains(stmt, query) {
+						return fmt.Errorf("injected autocommit error")
+					}
+					return nil
+				},
+			},
+		},
+	}
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(context.TODO())
+
+	sqlDB.SetMaxOpenConns(1)
+
+	if _, err := sqlDB.Exec(query); !testutils.IsError(err, "injected") {
+		t.Fatalf("expected injected error, got: %v", err)
+	}
+
+	var state string
+	if err := sqlDB.QueryRow("SHOW TRANSACTION STATUS").Scan(&state); err != nil {
+		t.Fatal(err)
+	}
+	if state != "NoTxn" {
+		t.Fatalf("expected state %s, got: %s", "NoTxn", state)
 	}
 }

--- a/pkg/sql/txn_restart_test.go
+++ b/pkg/sql/txn_restart_test.go
@@ -1728,3 +1728,75 @@ CREATE TABLE t.test (k INT PRIMARY KEY);
 			})
 	}
 }
+
+// Test that, if we'd otherwise perform an auto-retry but results for the
+// current txn have already been streamed to the client, we don't do the
+// auto-restart.
+func TestTxnAutoRetriesDisabledAfterResultsHaveBeenSentToClient(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	params, _ := createTestServerParams()
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(context.TODO())
+
+	tests := []struct {
+		name                  string
+		clientDirectedRetries bool
+	}{
+		{
+			name: "client_directed_retries",
+			clientDirectedRetries: true,
+		},
+		{
+			name: "no_client_directed_retries",
+			clientDirectedRetries: false,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Cleanup the connection state after each test so the next one can run
+			// statements.
+			// TODO(andrei): Once we're on go 1.9, this test should use the new
+			// db.Conn() method to tie each test to a connection; then this cleanup
+			// wouldn't be necessary. Also, the test is currently technically
+			// incorrect, as there's no guarantee that the state check at the end will
+			// happen on the right connection.
+			defer func() {
+				if _, err := sqlDB.Exec("ROLLBACK"); err != nil {
+					t.Fatal(err)
+				}
+			}()
+
+			var savepoint string
+			if tc.clientDirectedRetries {
+				savepoint = "SAVEPOINT cockroach_restart;"
+			}
+			// We'll run a statement that produces enough results to overflow the
+			// buffers and start streaming results to the client before the retriable
+			// error is injected.
+			sql := fmt.Sprintf(`
+				BEGIN TRANSACTION;
+				%s
+				SELECT generate_series(1, 10000);
+				SELECT crdb_internal.force_retry('1s');
+				COMMIT TRANSACTION;`, savepoint)
+			_, err := sqlDB.Exec(sql)
+			if !isRetryableErr(err) {
+				t.Fatalf("expected retriable error, got: %v", err)
+			}
+			var expectedState string
+			if tc.clientDirectedRetries {
+				expectedState = "RestartWait"
+			} else {
+				expectedState = "Aborted"
+			}
+			var state string
+			if err := sqlDB.QueryRow("SHOW TRANSACTION STATUS").Scan(&state); err != nil {
+				t.Fatal(err)
+			}
+			if state != expectedState {
+				t.Fatalf("expected state %s, got: %s", expectedState, state)
+			}
+		})
+	}
+}

--- a/pkg/sql/txn_restart_test.go
+++ b/pkg/sql/txn_restart_test.go
@@ -32,7 +32,6 @@ import (
 	"golang.org/x/net/context"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
-	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/server"
@@ -1015,7 +1014,7 @@ CREATE DATABASE t; CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
 	}
 }
 
-// Test that a COMMIT getting an error, retryable or not, leaves the txn
+// Test that a COMMIT getting an error, retriable or not, leaves the txn
 // finalized and not in Aborted/RestartWait (i.e. COMMIT, like ROLLBACK, is
 // always final). As opposed to an error on a COMMIT in an auto-retry
 // txn, where we retry the txn (not tested here).
@@ -1058,46 +1057,48 @@ CREATE DATABASE t; CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
 		{true},
 	}
 	for _, tc := range testCases {
-		const insertStmt = "INSERT INTO t.test(k, v) VALUES (0, 'boulanger')"
-		if err := aborter.QueueStmtForAbortion(
-			insertStmt, 1 /* abortCount */, false, /* willBeRetriedIbid */
-		); err != nil {
-			t.Fatal(err)
-		}
-		if _, err := sqlDB.Exec("BEGIN"); err != nil {
-			t.Fatal(err)
-		}
-		if tc.retryIntent {
-			if _, err := sqlDB.Exec("SAVEPOINT cockroach_restart"); err != nil {
+		t.Run(fmt.Sprintf("retryIntent=%t", tc.retryIntent), func(t *testing.T) {
+			const insertStmt = "INSERT INTO t.test(k, v) VALUES (0, 'boulanger')"
+			if err := aborter.QueueStmtForAbortion(
+				insertStmt, 1 /* abortCount */, false, /* willBeRetriedIbid */
+			); err != nil {
 				t.Fatal(err)
 			}
-		}
-		if _, err := sqlDB.Exec(insertStmt); err != nil {
-			t.Fatal(err)
-		}
-		if _, err := sqlDB.Exec("COMMIT"); !testutils.IsError(err, "pq: restart transaction") {
-			t.Fatalf("unexpected error: %v", err)
-		}
+			if _, err := sqlDB.Exec("BEGIN"); err != nil {
+				t.Fatal(err)
+			}
+			if tc.retryIntent {
+				if _, err := sqlDB.Exec("SAVEPOINT cockroach_restart"); err != nil {
+					t.Fatal(err)
+				}
+			}
+			if _, err := sqlDB.Exec(insertStmt); err != nil {
+				t.Fatal(err)
+			}
+			if _, err := sqlDB.Exec("COMMIT"); !testutils.IsError(err, "pq: restart transaction") {
+				t.Fatalf("unexpected error: %v", err)
+			}
 
-		// Check that we can start another txn on the (one and only) connection.
-		if _, err := sqlDB.Exec("BEGIN"); err != nil {
-			t.Fatal(err)
-		}
-		// Check that we don't see any rows, so the previous txn was rolled back.
-		rows, err := sqlDB.Query("SELECT * FROM t.test")
-		if err != nil {
-			t.Fatal(err)
-		}
-		defer rows.Close()
-		if rows.Next() {
-			var k int
-			var v string
-			err := rows.Scan(&k, &v)
-			t.Fatalf("found unexpected row: %d %s, %v", k, v, err)
-		}
-		if _, err := sqlDB.Exec("END"); err != nil {
-			t.Fatal(err)
-		}
+			// Check that we can start another txn on the (one and only) connection.
+			if _, err := sqlDB.Exec("BEGIN"); err != nil {
+				t.Fatal(err)
+			}
+			// Check that we don't see any rows, so the previous txn was rolled back.
+			rows, err := sqlDB.Query("SELECT * FROM t.test")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer rows.Close()
+			if rows.Next() {
+				var k int
+				var v string
+				err := rows.Scan(&k, &v)
+				t.Fatalf("found unexpected row: %d %s, %v", k, v, err)
+			}
+			if _, err := sqlDB.Exec("END"); err != nil {
+				t.Fatal(err)
+			}
+		})
 	}
 }
 
@@ -1250,29 +1251,6 @@ SELECT * from t.test WHERE k = 'test_key';
 	}
 	if !hitError {
 		t.Errorf("expected to hit error, but it didn't happen")
-	}
-}
-
-// TestNonRetryableErrorOnCommit verifies that a non-retryable error from the
-// execution of EndTransactionRequests is propagated to the client.
-func TestNonRetryableErrorOnCommit(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-
-	params, cmdFilters := createTestServerParams()
-	s, sqlDB, _ := serverutils.StartServer(t, params)
-	defer s.Stopper().Stop(context.TODO())
-
-	defer cmdFilters.AppendFilter(func(args storagebase.FilterArgs) *roachpb.Error {
-		if req, ok := args.Req.(*roachpb.EndTransactionRequest); ok {
-			if bytes.Contains(req.Key, []byte(keys.SystemConfigSpan.Key)) {
-				return roachpb.NewErrorWithTxn(errors.New(t.Name()), args.Hdr.Txn)
-			}
-		}
-		return nil
-	}, false)()
-
-	if _, err := sqlDB.Exec("CREATE DATABASE t"); !testutils.IsError(err, t.Name()) {
-		t.Errorf("unexpected error %v", err)
 	}
 }
 

